pkg/operator/ceph/cluster/osd/key_rotation.go

```diff
@@ -0,0 +1,279 @@
+*/
+
+package osd
+
+import (
+	"fmt"
+	"path/filepath"
+
+	"github.com/pkg/errors"
+	cephv1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
+	kms "github.com/rook/rook/pkg/daemon/ceph/osd/kms"
+	"github.com/rook/rook/pkg/operator/ceph/controller"
+	"github.com/rook/rook/pkg/operator/k8sutil"
+	batch "k8s.io/api/batch/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	ctrl "sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+const (
+	keyRotationCronJobAppName    = "rook-ceph-osd-key-rotation"
+	keyRotationCronJobAppNameFmt = "rook-ceph-osd-key-rotation-%d"
+)
+
+// getKeyRotationContainer returns the container spec for the key rotation job.
+func (c *Cluster) getKeyRotationContainer(osdProps osdProperties, volumeMounts []v1.VolumeMount, devices []string) (v1.Container, error) {
+	envVars := c.getConfigEnvVars(osdProps, k8sutil.DataDir, true)
+
+	// enable debug logging
+	envVars = append(envVars, setDebugLogLevelEnvVar(true))
+	envVars = append(envVars, v1.EnvVar{Name: "ROOK_CEPH_VERSION", Value: c.clusterInfo.CephVersion.CephVersionFormatted()})
+	envVars = append(envVars, kms.ConfigToEnvVar(c.spec)...)
+
+	// run privileged always since we always mount /dev
+	privileged := true
+	runAsUser := int64(0)
+	runAsNonRoot := false
+	readOnlyRootFilesystem := false
+
+	args := []string{"key-management", "rotate-key", osdProps.pvc.ClaimName}
+	args = append(args, devices...)
+
+	osdProvisionContainer := v1.Container{
+		Command:         []string{"rook"},
+		Args:            args,
+		Name:            keyRotationCronJobAppName,
+		Image:           c.rookVersion,
+		ImagePullPolicy: controller.GetContainerImagePullPolicy(c.spec.CephVersion.ImagePullPolicy),
+		VolumeMounts:    volumeMounts,
+		Env:             envVars,
+		EnvFrom:         getEnvFromSources(),
+		SecurityContext: &v1.SecurityContext{
+			Privileged:             &privileged,
+			RunAsUser:              &runAsUser,
+			RunAsNonRoot:           &runAsNonRoot,
+			ReadOnlyRootFilesystem: &readOnlyRootFilesystem,
+		},
+		Resources: osdProps.resources,
+	}
+
+	return osdProvisionContainer, nil
+}
+
+// applyKeyRotationPlacement applies the placement settings for the key rotation job
+// so that it is scheduled on the same node as the OSD.
+func (c *Cluster) applyKeyRotationPlacement(spec *v1.PodSpec, labels map[string]string) {
+	spec.TopologySpreadConstraints = nil
+	spec.Affinity.PodAntiAffinity = nil
+	spec.Affinity.PodAffinity = &v1.PodAffinity{
+		RequiredDuringSchedulingIgnoredDuringExecution: []v1.PodAffinityTerm{
+			{
+				LabelSelector: &metav1.LabelSelector{
+					MatchLabels: labels,
+				},
+				TopologyKey: "kubernetes.io/hostname",
+			},
+		},
+	}
+}
+
+// getKeyRotationPodTemplateSpec returns the pod template spec for the key rotation job.
+func (c *Cluster) getKeyRotationPodTemplateSpec(osdProps osdProperties, osd OSDInfo, restart v1.RestartPolicy) (*v1.PodTemplateSpec, error) {
```

Comment:
Some design questions:
- One cron job per OSD? 
- Could we instead have a single cron job per node that rotates all OSDs on that node? This would just require that OSDs not be portable unless the job could dynamically discover which OSDs on the node need to be rotated.
- Since we're using host paths, it will be critical to run the cron jobs on the same node where the OSDs are running. 
   - Is there a node selector? I'm not seeing how the cron job knows where to run.
   - If the OSD is portable and moves to another node, would we need a reconcile to run the job on the correct node? Or perhaps we can use pod affinity to the OSD?

Reply:
> Some design questions:
> 
> * One cron job per OSD?

yes

> * Could we instead have a single cron job per node that rotates all OSDs on that node? This would just require that OSDs not be portable unless the job could dynamically discover which OSDs on the node need to be rotated.
> * Since we're using host paths, it will be critical to run the cron jobs on the same node where the OSDs are running.

The cronjob per OSD will have podAffinity set with requiredDuringSchedule using that OSD's labels as selector, therefore the corresponding cronjob pod will always be sheduled on the node where its OSD is running.
Each CronJob will have access to only that OSD's underlying bridge host path where `block*-tmp` encrypted devices are mapped onto and can run only if that folder exists.